src/plugins/ui_actions/public/service/ui_actions_service.ts

```diff
@@ -69,15 +75,18 @@ export class UiActionsService {
     return trigger.contract;
   };
 
-  public readonly registerAction = (action: Action) => {
+  public readonly registerAction = <Context>(action: Action<Context>) => {
     if (this.actions.has(action.id)) {
       throw new Error(`Action [action.id = ${action.id}] already registered.`);
     }
 
     this.actions.set(action.id, action);
   };
 
-  public readonly attachAction = (triggerId: string, actionId: string): void => {
+  // TODO: make this
+  // <T extends TriggerId>(triggerId: T, action: Action<TriggerContextMapping[T]>): \
```

Comment:
Interesting! Have you checked if it works? Or it will be not that simple?

Reply:
It does seem to be working in my PR (still wip)!

<img width="1040" alt="Screen Shot 2020-02-27 at 11 53 34 AM" src="https://user-images.githubusercontent.com/16563603/75466188-d04dae00-5957-11ea-86f8-accebdb74a2d.png">